### ISSUE TITLE
Restart aborted before given a chance to complete, stuck in infinite loop

### ISSUE THREAD

**Issue by japicken** [rgl/packer-provisioner-windows-update#61]
Using the VSphereISO builder to create a Windows Server 2019 (1809) LTSC image, I'm seeing the following sequence of events, 100% reproducible in my environment.  The windows-update provisioner is getting stuck waiting in an infinite loop for a reboot that will never happen.

The relevant sequence of events (note timestamps):

```
2020/03/12 17:08:12 packer.exe plugin: [INFO] starting remote command: shutdown.exe -f -r -t 0 -c "packer restart"
2020/03/12 17:08:12 packer.exe plugin: [INFO] starting remote command: shutdown.exe -f -r -t 60 -c "packer restart test"
2020/03/12 17:08:13 packer.exe plugin: [INFO] starting remote command: shutdown.exe -a
```

It seems that the packer restart test and subsequent shutdown -a occur so quickly that the real, non-test, reboot is cancelled.

Full log:
```
    vsphere-iso: Waiting for operation to complete (system performance: 0% cpu; 23% memory)...
2020/03/12 17:08:12 packer.exe plugin: [INFO] command 'PowerShell -ExecutionPolicy Bypass -OutputFormat Text -File C:/Windows/Temp/packer-windows-update-elevated.ps1' exited with code: 101
2020/03/12 17:08:12 packer.exe plugin: [INFO] RPC endpoint: Communicator ended with: 101
==> vsphere-iso: Restarting the machine...
2020/03/12 17:08:12 [INFO] 1325 bytes written for 'stdout'
2020/03/12 17:08:12 [INFO] 0 bytes written for 'stderr'
2020/03/12 17:08:12 [INFO] RPC client: Communicator ended with: 101
2020/03/12 17:08:12 [INFO] RPC endpoint: Communicator ended with: 101
2020/03/12 17:08:12 packer-provisioner-windows-update.exe plugin: 2020/03/12 17:08:12 [INFO] 0 bytes written for 'stderr'
2020/03/12 17:08:12 packer-provisioner-windows-update.exe plugin: 2020/03/12 17:08:12 [INFO] 1325 bytes written for 'stdout'
2020/03/12 17:08:12 packer-provisioner-windows-update.exe plugin: 2020/03/12 17:08:12 [INFO] RPC client: Communicator ended with: 101
2020/03/12 17:08:12 packer.exe plugin: [INFO] starting remote command: shutdown.exe -f -r -t 0 -c "packer restart"
2020/03/12 17:08:12 packer.exe plugin: [INFO] command 'shutdown.exe -f -r -t 0 -c "packer restart"' exited with code: 0
2020/03/12 17:08:12 packer.exe plugin: [INFO] RPC endpoint: Communicator ended with: 0
2020/03/12 17:08:12 [INFO] 0 bytes written for 'stdout'
==> vsphere-iso: Waiting for machine to become available...
2020/03/12 17:08:12 [INFO] 0 bytes written for 'stderr'
2020/03/12 17:08:12 [INFO] RPC client: Communicator ended with: 0
2020/03/12 17:08:12 [INFO] RPC endpoint: Communicator ended with: 0
2020/03/12 17:08:12 packer-provisioner-windows-update.exe plugin: 2020/03/12 17:08:12 [INFO] 0 bytes written for 'stdout'
2020/03/12 17:08:12 packer-provisioner-windows-update.exe plugin: 2020/03/12 17:08:12 [INFO] 0 bytes written for 'stderr'
2020/03/12 17:08:12 packer-provisioner-windows-update.exe plugin: 2020/03/12 17:08:12 [INFO] RPC client: Communicator ended with: 0
2020/03/12 17:08:12 packer.exe plugin: [INFO] starting remote command: shutdown.exe -f -r -t 60 -c "packer restart test"
2020/03/12 17:08:12 packer.exe plugin: [INFO] command 'shutdown.exe -f -r -t 60 -c "packer restart test"' exited with code: 0
2020/03/12 17:08:12 packer.exe plugin: [INFO] RPC endpoint: Communicator ended with: 0
2020/03/12 17:08:12 [INFO] 0 bytes written for 'stdout'
2020/03/12 17:08:12 [INFO] 0 bytes written for 'stderr'
2020/03/12 17:08:12 [INFO] RPC client: Communicator ended with: 0
2020/03/12 17:08:12 [INFO] RPC endpoint: Communicator ended with: 0
2020/03/12 17:08:12 packer-provisioner-windows-update.exe plugin: 2020/03/12 17:08:12 [INFO] 0 bytes written for 'stderr'
2020/03/12 17:08:12 packer-provisioner-windows-update.exe plugin: 2020/03/12 17:08:12 [INFO] 0 bytes written for 'stdout'
2020/03/12 17:08:12 packer-provisioner-windows-update.exe plugin: 2020/03/12 17:08:12 [INFO] RPC client: Communicator ended with: 0
2020/03/12 17:08:13 packer.exe plugin: [INFO] starting remote command: shutdown.exe -a
2020/03/12 17:08:13 packer.exe plugin: [INFO] command 'shutdown.exe -a' exited with code: 0
2020/03/12 17:08:13 packer.exe plugin: [INFO] RPC endpoint: Communicator ended with: 0
2020/03/12 17:08:13 [INFO] 0 bytes written for 'stdout'
2020/03/12 17:08:13 [INFO] 0 bytes written for 'stderr'
2020/03/12 17:08:13 [INFO] RPC client: Communicator ended with: 0
2020/03/12 17:08:13 [INFO] RPC endpoint: Communicator ended with: 0
2020/03/12 17:08:13 packer-provisioner-windows-update.exe plugin: 2020/03/12 17:08:13 [INFO] 0 bytes written for 'stderr'
2020/03/12 17:08:13 packer-provisioner-windows-update.exe plugin: 2020/03/12 17:08:13 [INFO] 0 bytes written for 'stdout'
2020/03/12 17:08:13 packer-provisioner-windows-update.exe plugin: 2020/03/12 17:08:13 [INFO] RPC client: Communicator ended with: 0
2020/03/12 17:08:13 packer.exe plugin: [INFO] starting remote command: PowerShell -ExecutionPolicy Bypass -OutputFormat Text -File C:/Windows/Temp/packer-windows-update-pending-reboot-elevated.ps1
    vsphere-iso: Waiting for the Windows Modules Installer to exit...
2020/03/12 17:08:39 packer.exe plugin: [INFO] command 'PowerShell -ExecutionPolicy Bypass -OutputFormat Text -File C:/Windows/Temp/packer-windows-update-pending-reboot-elevated.ps1' exited with code: 101
2020/03/12 17:08:39 packer.exe plugin: [INFO] RPC endpoint: Communicator ended with: 101
2020/03/12 17:08:39 [INFO] 0 bytes written for 'stderr'
2020/03/12 17:08:39 [INFO] 54 bytes written for 'stdout'
2020/03/12 17:08:39 [INFO] RPC client: Communicator ended with: 101
2020/03/12 17:08:39 [INFO] RPC endpoint: Communicator ended with: 101
2020/03/12 17:08:39 packer-provisioner-windows-update.exe plugin: 2020/03/12 17:08:39 [INFO] 0 bytes written for 'stderr'
2020/03/12 17:08:39 packer-provisioner-windows-update.exe plugin: 2020/03/12 17:08:39 [INFO] 54 bytes written for 'stdout'
2020/03/12 17:08:39 packer-provisioner-windows-update.exe plugin: 2020/03/12 17:08:39 [INFO] RPC client: Communicator ended with: 101
2020/03/12 17:08:39 packer-provisioner-windows-update.exe plugin: 2020/03/12 17:08:39 Retryable error: Machine not yet available (exit status 101)
2020/03/12 17:08:44 packer.exe plugin: [INFO] starting remote command: shutdown.exe -f -r -t 60 -c "packer restart test"
2020/03/12 17:08:44 packer.exe plugin: [INFO] command 'shutdown.exe -f -r -t 60 -c "packer restart test"' exited with code: 0
2020/03/12 17:08:44 packer.exe plugin: [INFO] RPC endpoint: Communicator ended with: 0
2020/03/12 17:08:44 [INFO] 0 bytes written for 'stderr'
2020/03/12 17:08:44 [INFO] 0 bytes written for 'stdout'
2020/03/12 17:08:44 [INFO] RPC client: Communicator ended with: 0
2020/03/12 17:08:44 [INFO] RPC endpoint: Communicator ended with: 0
2020/03/12 17:08:44 packer-provisioner-windows-update.exe plugin: 2020/03/12 17:08:44 [INFO] 0 bytes written for 'stderr'
2020/03/12 17:08:44 packer-provisioner-windows-update.exe plugin: 2020/03/12 17:08:44 [INFO] 0 bytes written for 'stdout'
2020/03/12 17:08:44 packer-provisioner-windows-update.exe plugin: 2020/03/12 17:08:44 [INFO] RPC client: Communicator ended with: 0
2020/03/12 17:08:44 packer.exe plugin: [INFO] starting remote command: shutdown.exe -a
2020/03/12 17:08:44 packer.exe plugin: [INFO] command 'shutdown.exe -a' exited with code: 0
2020/03/12 17:08:44 packer.exe plugin: [INFO] RPC endpoint: Communicator ended with: 0
2020/03/12 17:08:44 [INFO] 0 bytes written for 'stderr'
2020/03/12 17:08:44 [INFO] 0 bytes written for 'stdout'
2020/03/12 17:08:44 [INFO] RPC client: Communicator ended with: 0
2020/03/12 17:08:44 [INFO] RPC endpoint: Communicator ended with: 0
2020/03/12 17:08:44 packer-provisioner-windows-update.exe plugin: 2020/03/12 17:08:44 [INFO] 0 bytes written for 'stderr'
2020/03/12 17:08:44 packer-provisioner-windows-update.exe plugin: 2020/03/12 17:08:44 [INFO] 0 bytes written for 'stdout'
2020/03/12 17:08:44 packer-provisioner-windows-update.exe plugin: 2020/03/12 17:08:44 [INFO] RPC client: Communicator ended with: 0
2020/03/12 17:08:44 packer.exe plugin: [INFO] starting remote command: PowerShell -ExecutionPolicy Bypass -OutputFormat Text -File C:/Windows/Temp/packer-windows-update-pending-reboot-elevated.ps1
    vsphere-iso: Waiting for the Windows Modules Installer to exit...

// And on and on forever...
```

**Comment by japicken** [comment]
I have a functional, but not optimal fix for this locally, but this would be the first Go I've ever touched.  The root cause is restartCommand is reporting success, but failing to do it's job (thank you Windows).  Another contributing cause is that when this happens, we never get around to retrying the restartCommand, because we're caught in the next retry loop (in `func (p *Provisioner) restart`).

The retry loop we are caught in is retrying too much, in other words, if testRestartCommand and abortTestRestartCommand report success (which they are in the failure scenario), then pendingRebootElevatedCommand will never succeed because no reboot is ever happening.

So the fix is basically if testRestartCommand and abortRestartCommand succeed, but pendingRebootElevatedCommand fails, we should try to reboot again.

**Comment by rgl** [comment]
Thanks for looking into this!
I have a couple of notes:
* the exit code of remote execution sometimes is fake because something in the stack is failing to report there was a communication failure at the socket/network layer. This means we should send a known string alike "exiting with exit code y" and parse that in the go code.
* the reboot logic was borrowed from the reboot logic of packer, so maybe they have improved it in the mean time.
*

**Comment by japicken** [comment]
I did some looking into your couple of notes.
- Not sure how to achieve this.  I could not find the code to fix the underlying issue.  Instead I made the restart function robust against this failure by testing that the machine no longer has a pendingReboot.
- The merge from the place where this was borrowed was more work than I can put in here, and I'm not entirely sure it's resolved.  Difficult for me to do this without a common ancestor or deep knowledge of the components involved.

Thanks for your help and making this provisioner for all of us to use!
